### PR TITLE
Fix incorrect fluxes and uncertainties returned by FluxConservingResampler, increase computation speed 

### PR DESCRIPTION
**Pull request by cshanahan1**
(Resolves issue https://github.com/astropy/specutils/issues/817, https://github.com/astropy/specutils/issues/1059)

As explained in #1059, the current implementation of FluxConservingResampler does not do the exact method described in https://arxiv.org/abs/1705.05165. The discrepancy becomes apparent in cases of partial overlapping bins with uneven bin sizes. This PR fixes that, and will return new (correct) fluxes and errors in those cases. For this reason, some of the 'truth' values in the tests are changed here.

Additionally, this fixes the issue described in #817 where the computation time is very long (especially in cases like the example there, resampling ~a million points down to a few hundred or thousand). With the new approach in this PR, the time to resample decreases from 4 minutes to about half a second. The following plot and table shows the speed comparison for the current version vs. this new one.
compare

| n_bins | old time | new time |
| --- | --- | --- |
| 2 |      0.0488 | 0.0046 |
| 3 |       0.0463 |  0.1264 |
| 20 |      0.291 |  0.2337 |
| 50  |     0.7302 |  0.2456 |
| 100  |    2.8603 |  0.2528 |
| 250  |    6.9634 |  0.253 |
| 500  |    14.6813 |  0.2578 |
| 1000  |   30.487 |  0.2655 |
| 2000  |   94.8554 |  0.3083 |
| 3000  |   255.93 |  0.4516 |

I do want to note that there is a small domain where the matrix approach is faster (a million down to 2-20 samples in that case, see table). We could consider eventually also supporting a matrix operation approach in addition to this method, and determining based on the input which should be faster, but for now I think this new version represents a significant improvement for most cases.

To summarize the changes, the current approach constructs a (n_input_samples) * (n_output_samples) array, where each matrix element (ij) represents the overlap between an input (i) and output (j) bin. Therefore, these matrices are huge with most elements 0.

The new approach calculates the new fluxes and errors iteratively (looping over output wavelengths), comparing the minimal number of input wavelengths possible. It starts by clipping out any bins from the output wavelength array that go out of range of the input wavelengths, and then another clip again on the input bin array after figuring out the minimal overlapping bin. From there, it will compare bins until the front edge of an output bin exceeds the final edge of that output bin. If an input bin terminates within an output bin, it will be excluded in the comparison next time, insuring there are no redundant comparisons. 

(I closed the first PR for this https://github.com/astropy/specutils/pull/1058)

### CODE REVIEW

**Comment by pllim** [comment]
The timing stated in the table is very impressive. Good job! 👏 

Does @sean-lockwood want to take this PR branch for a spin with his original use case?

> I do want to note that there is a small domain where the matrix approach is faster (a million down to 2-20 samples in that case, see table)

I don't think `0.05 s -> 0.1 s` for that narrow range of samples is worth re-introducing matrix.

File: specutils/manipulation/resample.py
Comment (rosteen) [inline specutils/manipulation/resample.py:261]: I think this is the only thing in this implementation that worries me - the previous one handled any higher dimensionality for flux, and I think that the most common cases are 1D and 3D. Could you generalize this at least to 3D?

Comment (cshanahan1) [reply specutils/manipulation/resample.py]: done!

File: CHANGES.rst
Comment (eteq) [inline CHANGES.rst:10]:
```suggestion
- Reimplementation of FluxConservingResampler. It is now faster and yields more accurate results. [#1060]
```
Rewording this to be more "user facing" - i.e., it tells what you get from this PR.

File: specutils/manipulation/resample.py
Comment (eteq) [inline specutils/manipulation/resample.py:116]: Is it actually necessary to use `.value` here? It doesn't matter too much given the conversion to match units has already happened, but if it doesn't have a negative consequence I think it's safer not to use `.value` since there's always the chance that this will catch some unit conversion issues that slip through.

**Review by eteq — APPROVED** [review]
Looking good!  I have a couple of take-it-or-leave-it-as-you-wish inline comments, @cshanahan1, and a few broader items, but all are optional some I'm approving this meaning "you can either make these changes or not and I'm ok with this either way".

General items:
 1) https://arxiv.org/abs/1705.05165 mentions the `spectres` backage.  I also mentioned this in https://github.com/astropy/specutils/issues/817#issuecomment-1497776023 but as @larrybradley pointed out in https://github.com/astropy/specutils/issues/817#issuecomment-1516745403 it is license-incompatible, so we cannot use the `spectres` code.  *however*, that doesn't stop us using `spectres` to build a test-case for this code (see https://www.gnu.org/licenses/gpl-faq.en.html#GPLOutput which highlights that the GPL cannot be applied to outputs of programs).  So I would encourage adding a test where you generate an *output* spectrum from spectres and test that this new version matches that. This will hopefully ensure that in the future we don't accidentally make breaking changes that violate the algorithm like what clearly happened originally here.

2) This is probably not a suggestion for *this* PR, rather a possible solution for the point you made that the matrix version is faster for small datasets: consider instead implementing this algorithm as a *sparse* matrix operation.  As I mentioned in https://github.com/astropy/specutils/issues/817#issuecomment-1497776023 I think that's the flaw in the original implementation - it's O(n^2) because almost all the time is spent multiplying zeros by zeros, and that's *exactly* the reason why sparse matrix libraries exist, and I suspect it will end up a lot faster than this for-loop based implementation if we can recast it that way.  But given this algorithm is still way way better than the previous one, this is definitely "future work" and shouldn't block this PR.

3) as an alternative to the above - a similar "how to speed it up for small n or other cases" approach might be to use numba or one of the other python jitting tools.  Those work especially well in cases like this where there's a for loop over an array, so it might be an easy way to get a good speedup.  Again, though, no need to block this PR unless you think you can do it really easily and quickly.

File: specutils/manipulation/resample.py
Comment (cshanahan1) [reply specutils/manipulation/resample.py:116]: shouldn't be necessary, i was getting a bunch of TypeErrors i didn't want to track down haha. i'll look into it

Comment (cshanahan1) [reply specutils/manipulation/resample.py]: Done!

**Comment by cshanahan1** [comment]
Thanks for the review @eteq. I added a new test that takes care of item 1 on your list of suggestions - is that what you had in mind? I ran the code from the other package to generate output for the test rather than adding it as a test dependency and generating it each time.

As for items 2 and 3, I think those are good ideas and I will make follow up tickets for them but I want to get this PR merged first.

**Comment by sean-lockwood** [comment]
I ran the calculations again, so the numbers won't exactly match the plots above.  Values are in seconds.

```
# out      1.10.0       0.1.dev2966+g7c71c41

 200     2.145341       0.179401
 250     2.625112       0.144806
 300     3.138709       0.150052
 350     3.728590       0.147948
 400     4.217818       0.147381
 450     4.709783       0.147746
 500     5.271168       0.152145
 550     5.813444       0.150349
 600     6.293068       0.156625
 650     6.798396       0.151181
 700     7.413938       0.151452
1200    12.702604       0.157407
1700    29.881622       0.164718
2200    50.650592       0.171427
2700    73.037056       0.181266
```

**Review by rosteen — APPROVED** [review]
Thank you for all the work on this, this is a great improvement. I'll hold off on hitting the merge button until tomorrow to see the answer to @pllim's last question (although I think that concern was addressed by checking against the `specres` output).

**Comment by sean-lockwood** [comment]
Ah, I understand.  I just took a quick look at the output flux.  Most points do well, but a handful have up to 30% discrepancy in flux.  I would have to do further testing to understand the origin of the difference.

flux_comparison

**Comment by cshanahan1** [comment]
@sean-lockwood if you can send me your input spectral axis, flux, and err values i can compare it against Spectres to make sure those differences are expected

**Comment by sean-lockwood** [comment]
The uncertainty array was None in my test.

Input and outputs here:
[spec_files.zip](https://github.com/astropy/specutils/files/11617745/spec_files.zip)

Read output arrays via:
```python
import pickle

with open('v1.10.0_downsampled.pkl', 'rb') as f:
    v1_wave, v1_flux = pickle.load(f)

with open('new_downsampled.pkl', 'rb') as f:
    new_wave, new_flux = pickle.load(f)
```

Generated via:
```python
import astropy.units as u
import specutils
from specutils import Spectrum1D
from specutils.manipulation import FluxConservingResampler

print(specutils.__version__)

orig = Spectrum1D.read('lte012.0-3.5-0.0a+0.0.BT-Settl.spec.fits')

resampler = FluxConservingResampler()
output_wavelengths = np.linspace(12.58e-4, 776247.11e-4, num=1500) * u.um
resampled_spec = resampler(orig, output_wavelengths)
```

**Comment by cshanahan1** [comment]
the bad news is that I spent too much time chasing down a non-issue, but the good news is that i determined it was a non-issue: spectres and this new version of FluxConservingResampler match basically within floating point error. (my dumb mistake and what made me worried was that i saw 0.2-0.3 offsets and interpreted these as percent differences when they were in fact absolute differences). 

it makes sense that @sean-lockwood was seeing a general agreement but an occasional 20-30% difference between the old version and the new version - those were the cases of overlapping input+output bins that were previously being treated incorrectly that i described in the linked issue. those few instances can be seen here:
Screen Shot 2023-06-01 at 11 38 43 PM


what matters is that spectres and this new version agree, and they do:

Screen Shot 2023-06-01 at 11 36 53 PM

**Comment by pllim** [comment]
How big is this "lte012.0-3.5-0.0a+0.0.BT-Settl.spec.fits" file? Maybe we should turn this Spectre use case into a unit test (as a follow-up PR) to prevent regression.

**Review by pllim — APPROVED** [review]
Based on Sean's response and the subsequent finding, I am satisfied with both performance improvement and bug fix.

Thank you very much, everyone! And excellent work, @cshanahan1 !

**Comment by sean-lockwood** [comment]
Attaching a truncated version of the input file (now 847k uncompressed):
[truncated_lte012.0-3.5-0.0a+0.0.BT-Settl.spec.fits.zip](https://github.com/astropy/specutils/files/11636837/truncated_lte012.0-3.5-0.0a%2B0.0.BT-Settl.spec.fits.zip)

Still gives the discrepant point:
truncated_comparison

**Comment by cshanahan1** [comment]
@sean-lockwood it makes sense there are some discrepant points when comparing the new version vs the old - this PR also includes a fix for some instances where fluxes and errors were being calculated incorrectly (when input and output bins overlap in a certain way).

**Comment by sean-lockwood** [comment]
@cshanahan1 
Makes sense to me.  Was just attaching a smaller version of the file for the unit test.

**Comment by pllim** [comment]
Thanks! I created https://github.com/astropy/specutils/issues/1064 to add a test, for whoever who wants to tackle it.

**Pull request merged**
